pkg/log/logger.go

```diff
@@ -0,0 +1,298 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package log
+
+import (
+	"fmt"
+	"math"
+	"os"
+)
+
+// Level describes the severity of log messages.
+type Level int
+
+const (
+	// LevelDebug is the severity for debug messages.
+	LevelDebug Level = iota
+	// LevelInfo is the severity for informational messages.
+	LevelInfo
+	// LevelWarn is the severity for warnings.
+	LevelWarn
+	// LevelError is the severity for errors.
+	LevelError
+	// LevelFatal is the severity for fatal errors.
+	LevelFatal
+	// LevelPanic is the severity for panic messages.
+	LevelPanic
+)
+
+// Logger is the interface for producing log messages for/from a particular source.
+type Logger interface {
+	// Debug formats and emits a debug message.
+	Debug(format string, args ...interface{})
+	// Info formats and emits an informational message.
+	Info(format string, args ...interface{})
+	// Warn formats and emits a warning message.
+	Warn(format string, args ...interface{})
+	// Error formats and emits an error message.
+	Error(format string, args ...interface{})
+	// Fatal formats and emits an error message and os.Exit()'s with status 1.
+	Fatal(format string, args ...interface{})
+	// Panic formats and emits an error messages, and panics with the same.
+	Panic(format string, args ...interface{})
+
+	// DebugBlock formats and emits a multiline debug message.
+	DebugBlock(prefix string, format string, args ...interface{})
+	// InfoBlock formats and emits a multiline information message.
+	InfoBlock(prefix string, format string, args ...interface{})
+	// WarnBlock formats and emits a multiline warning message.
+	WarnBlock(prefix string, format string, args ...interface{})
+	// ErrorBlock formats and emits a multiline error message.
+	ErrorBlock(prefix string, format string, args ...interface{})
+
+	// EnableDebug enables debug messages for this Logger.
+	EnableDebug(bool) bool
+	// DebugEnabled checks if debug messages are enabled for this Logger.
+	DebugEnabled() bool
+
+	// Source returns the source name of this Logger.
+	Source() string
+}
+
+// logger implements our Logger.
+type logger uint
+
+// EnableDebug enables/disables debug logging for this logger.
+func (l logger) EnableDebug(state bool) bool {
+	log.Lock()
+	defer log.Unlock()
+
+	cfg, _ := log.configs[l]
+	old := cfg.setTracing(state)
+	log.configs[l] = cfg
+
+	return old
+}
+
+// DebugEnabled checks debug logging is enabled for this logger.
+func (l logger) DebugEnabled() bool {
+	log.RLock()
+	defer log.RUnlock()
+
+	cfg, _ := log.configs[l]
+
+	return cfg.isTracing()
+}
+
+// Source returns the source for the given logger.
+func (l logger) Source() string {
+	log.RLock()
+	defer log.RUnlock()
+
+	source, _ := log.sources[l]
+
+	return source
+}
+
+// Debug logs a debug message.
+func (l logger) Debug(format string, args ...interface{}) {
```

Comment:
There could be one backend function for all these, maybe something like

```
func (l logger) logMessage(level Level, format string, args ..interface{}) {
	if cfg, active, emit := l.config(level); emit {
		active.Log(level, cfg.source(), format, args...)
	}
}
```

Reply:
The lack of any kind of macros is a nuisance in golang. Here is a primary example...

For debug messages, I want to drop the suppressed ones as early as possible. Before they cross any potential goroutine boundary, IOW gets injected into a channel, then eventually pulled out at the other end by the backend just to get dropped, just like with the default fmt backend.

For the others it's true we could combine them to a single function. However, since we don't have macros I don't see a reasonable way of keeping the API like it is, yet combining Info, Warn, Error, Fatal, and Panic into a single function, while leaving Debug a separate one.